### ISSUE TITLE
Unable to convert image pixels to viewport coordinates using imageToViewerElementCoordinates/ imageToViewportCoordinates

### ISSUE THREAD

**Issue by IrisLiuCY** [openseadragon/openseadragon#1338]
I am trying to convert my coordinates from image-pixel representation to viewport coordinates so that I can create overlay shapes on OSD using openseadragon fabric js plugin. As far as I know, fabric js plugin uses viewport coordinates and so does OSD.  

However, for some reasons, the function `imageToViewerElementCoordinates/ imageToViewportCoordinates` does not help to transform my image pixel to viewport coordinates. Below you can see my codes:
```

imagePolyPts[j] = new OpenSeadragon.Point (polyPts[j].x,polyPts[j].y);
console.log("imagePolyPts", imagePolyPts[j]);
viewportPoint[j] = viewer.viewport.imageToViewerElementCoordinates(imagePolyPts[j]); 
console.log("viewportPoint", viewportPoint[j]);
```

And the result output in the console is as below. You can see the two values are the same. Anyone knows the reason for that?
![image2viewport](https://user-images.githubusercontent.com/15152847/32324088-1abee9ca-c005-11e7-9407-aa4bdd80d901.png)

**Comment by iangilman** [comment]
`imageToViewerElementCoordinates` converts to the coordinates of the web page, not the viewport coordinates. I suppose the numbers would match if you had zoomed in so the image pixels were 1:1 with the web page pixels.

You want `imageToViewportCoordinates`; that should give different coordinates (unless you laid your image out 1:1 in the world like I suggested in https://github.com/openseadragon/openseadragon/issues/1220). Speaking of which, if you make your image coordinates and viewport coordinates the same, you don't need to do any conversion.

**Comment by IrisLiuCY** [comment]
Thanks for the reply. Unfortunately, I did try both methods 1) suggested as in #1220 or 2) use the `imageToViewerElementCoordinates` to convert the coordinates but neither of these produces the expected outcomes. 

You can see my settings as below:
`var picture = {
                  Image: {
                      xmlns: "http://schemas.microsoft.com/deepzoom/2008",
                      Url: "./Tumor_011_files/",
                      Format: "jpeg",
                      Overlap: "1",
                      TileSize: "254",
                      Size: {
                          Width:  "97792",
                          Height: "221184"
                      }
                  },
                  Width: 97792
              };

              this.viewer = OpenSeadragon({
                  id: "seadragon-viewer",
                  prefixUrl: "./openseadragon/images/", 
                  showNavigator: true,
                  gestureSettingsMouse:   {
                      clickToZoom: false
                  },
                  clickToZoom: "false",
                  defaultZoomLevel: 0, 
                  minZoomLevel: 0,
                  maxZoomLevel: 10,
                  visibilityRatio: 1.0,
                  constrainDuringPan: true, 
                  debugMode: false,
                  tileSources: picture
              });`

Of course, when I set the image as 1:1 settings, I didn't use the `imageToViewportCoordinates` to do conversion. One thing to note is that I use fabricjs overlay libraries. As I set OSD to image as 1:1, I also initialize the overlay as 

`var options = {
                  scale: 97792
              };
              var overlay = viewer.fabricjsOverlay(options);
`
Do you thing there are other possible bugs that I have missed? Please let me know.

**Comment by iangilman** [comment]
You're not actually setting the width properly... I know the correct format is confusing. Just leave your tileSource alone like so:

```
var picture = {
    Image: {
        xmlns: "http://schemas.microsoft.com/deepzoom/2008",
        Url: "./Tumor_011_files/",
        Format: "jpeg",
        Overlap: "1",
        TileSize: "254",
        Size: {
            Width:  "97792",
            Height: "221184"
        }
    }
};
```

...and set the width like this:

```
this.viewer = OpenSeadragon({
    ...
    tileSources: {
        tileSource: picture,
        width: 97792
    } 
});
```

Let me know if that works!

**Comment by IrisLiuCY** [comment]
Thanks iangilman, it works wonder! The viewport now stretched to 1:1 as the image's width. 
Below is the result of the viewport coordinates: 
<img width="232" alt="screen shot 2017-11-07 at 11 46 28 am" src="https://user-images.githubusercontent.com/15152847/32476298-5f9a7726-c3b1-11e7-9bc5-9029d7ea9a70.png">

**Comment by iangilman** [comment]
Great!

**Comment by tofetpuzo** [comment]
@iangilman This is a follow up question from above, so I have some coordinates say (x, y) -> (500, 500) and some ROI(Region of interest) labeled 1, 2, 3 -> , each roi maps to this (x, y) coordinates and corresponds to a small image on the tiledImage. I wan to be able to specify this (x,y) coordinates and want my cursor to zoom to that roi, how do I do it?

**Comment by iangilman** [comment]
Assuming you're using the technique above to map image pixels to viewport pixels, you could use `fitBounds` with the coordinates you want. For instance, if you want to show `w` pixels wide and `h` pixels tall centered on the (x,y) point, you could do:  

```
viewer.viewport.fitBounds(OpenSeadragon.Rect(x - w / 2, y - h / 2, w, h));
```

**Comment by tofetpuzo** [comment]
@iangilman Thank you so much for your response, please look at this code tell me why the dropdown is not firing, I am lost do I need a addHandler function(with 'open')  to make the image zoomTo or PanTo the coordinates. I just want to be able to zoom to an image once I click a dropdown(1, 2, 3) from the return

```
useEffect(() => { 
        window.viewer = viewer
        function ROICoordinatesTrigger(view_exported_roi_coordinates, event) {
            console.log("I am getting here")
        
            if(viewer && view_exported_roi_coordinates){
                // viewer.viewport.fitBounds(OpenSeadragon.Rect(x - w / 2, y - h / 2, w, h));
                const [x, y] = view_exported_roi_coordinates          
                var imagePolyPts = new OpenSeaDragon.Point(x, y);
                console.log(imagePolyPts)
                viewer.viewport.panTo(imagePolyPts, true)
                return viewer.viewport.zoomTo(3)
            }
            // viewer.setFullScreen(false);
        }
       ROICoordinatesTrigger(selectedROICoordinates)
       console.log("----")      
    }, [viewer, selectedROICoordinates])


return (
    <div id="roi-picker">
    <select id="roi-picker" onChange={handleROIDropdown}
    style={{ height: "20px", width: "70px", position: "absolute", top: "50", left: "50" }}>
        {data && Object.entries(data).map(([roi, coordinates]) => (<option key={roi} value={coordinates}>{roi}</option>))}
    </select>
    </div>
      );            
};
```

**Comment by iangilman** [comment]
It looks reasonable to me. Is the `console.log(imagePolyPts)` firing? What numbers are you seeing there?

What happens if you `console.log(viewer.world.getHomeBounds())`? How does it compare to the `imagePolyPts`?

**Comment by tofetpuzo** [comment]
@iangilman yes console.log(imagePolyPts) is firing but viewer.ZoomTo or PanTo does not trigger at all :( 

here are the information you requested
for console.log(viewer.world.getHomeBounds()) 
```
__webpack_modules__../node_modules/openseadragon/build/openseadragon/openseadragon.js.$.Rect {x: 0, y: 0, width: 1, height: 1.9260554584061953, degrees: 0} degrees: 0 height: 1.9260554584061953
width: 1 x: 0 y: 0 [[Prototype]]
```
console.log(imagePolyPts) : This x and y values was what I supplied from my api call which gives roi and its respective coordinates. roi(1, 2, 3 ... 160) then corresponding coordinates (x: 2173 y: 632, ...), All I want OSD to do is zoom to this coordinates. when I click on the select element in html which has an onChange parameter. So pass the function and zoom to that location.
```
__webpack_modules__../node_modules/openseadragon/build/openseadragon/openseadragon.js.$.Point {x: 2173, y: 632}
x: 2173 y: 632

```


Please can I just trouble you a bit, can you give me a sample of code where if I want to zoom and pan to a particular coordinates when I have a drop down(select html properties ) with different numbers(roi)(1.2,3) once I select this dropdown maybe 2 OSD zooms to that particular coordinates.

**Comment by iangilman** [comment]
Okay, you can see from the `getHomeBounds` that the width of the content (the image) in viewport coordinates is 1, and the height is 1.9260554584061953. The numbers you're using for pan are things like 2173, 632, which is way outside of that area! I believe that's why it's not working. 

The [discussion near the top](https://github.com/openseadragon/openseadragon/issues/1338#issuecomment-342230680) of this issue is about how to set up your image so the viewport coordinates match the image coordinates, which makes these calculations easier. For instance, if your image is 5000 pixels wide, you could do it like this:

```
    tileSources: {
        tileSource: 'image.dzi',
        width: 5000
    } 
```

There are other ways to translate between viewport and image coordinates, but that's a pretty convenient one. For more info on the different coordinate systems, see: https://openseadragon.github.io/examples/viewport-coordinates/

Other than that, the code you gave seems like it should work.

**Comment by abhinav2712** [comment]
hey @iangilman , lets say i have one bounding box on a normal image, and I want to draw similar annotation using annotorious of osd on the OSD viewer of same tiled image, how can I match the annotation size and postion with the bounding box? i used `imageToViewportCoordinates` but doesnt seem to work

**Comment by iangilman** [comment]
@abhinav2712 Looks like maybe your question has been answered on Discord?

https://discord.com/channels/888321246431969280/888820630865186876/1212759318366134272
